selfdrive/boardd/panda_flashing.cc

```diff
@@ -0,0 +1,212 @@
+  // Not sure how to copy from string to unsigned char * with one command. For loop works
+  for (int i = 0 ; i < program.length() ; i++) {
+    padded_program[i] = program[i];
+  }
+  for (int i = 0 ; i < paddedLength/ blockSize ; i++) {
+    LOGD("Programming with block %d", i);
+    dfuPanda->control_write(0x21, DFU_DNLOAD, 2 + i, 0, padded_program + blockSize * i, blockSize);
+    dfu_status(dfuPanda);
+  }
+  LOGD("Done with programming");
+}
+
+void dfu_reset(PandaComm* dfuPanda) {
+  unsigned char data[5];
+  data[0] = 0x21;
+  int adress = 0x8000000;
+  memcpy(&data[1], &adress, sizeof(adress));
+  dfuPanda->control_write(0x21, DFU_DNLOAD, 0, 0, data, 5);
+  dfu_status(dfuPanda);
+  try {
+    dfuPanda->control_write(0x21, DFU_DNLOAD, 2, 0, nullptr, 0);
+    unsigned char buf[6];
+    dfuPanda->control_read(0x21, DFU_GETSTATUS, 0, 0, buf, 6);
+  } catch(std::runtime_error &e) {
+    LOGE("DFU reset failed");
+  }
+}
+
+void dfu_program_bootstub(PandaComm* dfuPanda, std::string program) {
+   dfu_clear_status(dfuPanda);
+   dfu_erase(dfuPanda, 0x8004000);
+   dfu_erase(dfuPanda, 0x8000000);
+   dfu_program(dfuPanda, 0x8000000, program);
+   dfu_reset(dfuPanda);
+}
+
+void dfu_recover(PandaComm* dfuPanda) {
+  build_st("obj/bootstub.panda.bin");
+  std::string program = util::read_file(basedir + "/panda/board/obj/bootstub.panda.bin");
+  dfu_program_bootstub(dfuPanda, program);
+}
+
+void get_out_of_dfu(std::string dfu_serial) {
+  PandaComm* dfuPanda;
+  try {
+    dfuPanda = new PandaComm(0x0483, 0xdf11, dfu_serial);
+  } catch(std::runtime_error &e) {
+    LOGD("DFU panda not found");
+    return;
+  }
+  LOGD("Panda in DFU mode found, flashing recovery");
+  dfu_recover(dfuPanda);
+  delete(dfuPanda);
+}
+
+void update_panda(std::string serial, std::string dfu_serial) {
+  LOGD("updating panda");
+  LOGD("\n1: Move out of DFU\n");
+  get_out_of_dfu(dfu_serial);
+  LOGD("\n2: Start DynamicPanda and run the required steps\n");
+
+  std::string fw_fn = get_firmware_fn();
+  std::string fw_signature = get_expected_signature();
+  DynamicPanda tempPanda(serial, dfu_serial);
+  std::string panda_signature = tempPanda.bootstub ? "": tempPanda.get_signature();
+  LOGD("fw_sig::panda_sig");
+  LOGD(fw_signature.c_str());
+  LOGD(panda_signature.c_str());
+
+  if (tempPanda.bootstub || panda_signature != fw_signature) {
+    LOGW("Panda firmware out of date, update required");
+    tempPanda.flash(fw_fn);
+    LOGD("Done flashing new firmware");
+  }
+
+  if (tempPanda.bootstub) {
+    std::string bootstub_version = tempPanda.get_version();
+    LOGW("Flashed firmware not booting, flashing development bootloader. Bootstub verstion: ");
+    LOGW(bootstub_version.c_str());
+    tempPanda.recover();
+    LOGD("Done flashing dev bootloader and firmware");
+  }
+
+  if (tempPanda.bootstub) {
+    LOGW("Panda still not booting, exiting");
+    throw std::runtime_error("PANDA NOT BOOTING");
```

Comment:
Don't just throw exceptions. If running under boardd we want to keep retrying forever (with some delay)